### DISCUSSION TITLE
Asynchronous loading of messages with notmuch queries

### DISCUSSION THREAD

**Discussion by pasha009** [neomutt/neomutt#4501] [Development]
I am using Notmuch feature of neomutt. 
Here is relevant code in my muttrc:
```
# This variable specifies notmuch query type, supported types: 'threads' and
# 'messages'.
set nm_query_type = messages

# When set, mutt will use the first virtual mailbox (see virtual-mailboxes)
# as a spoolfile.
set virtual_spoolfile = yes

# virtual-mailboxes description notmuch-URI { description notmuch-URI ...}
source mailboxes

# notmuch keybindings
macro index,pager A "<modify-labels-then-hide>+archive -unread -inbox" "Archive message"
macro index,pager c "<change-vfolder>?" "Change to vfolder overview"
macro index,pager \\ "<vfolder-from-query>" "Search mailbox"
macro index,pager d "<modify-labels> +deleted -inbox -archive<enter> <delete-message>" "Delete Message" 
```

I am seeing a delay in opening virtual folders and in searching. It seems that it is the delay of finding all mails of the search term or mailbox and then sorting them.

_Is there some way such that one can instantly see partial search result and rest of the result appends asynchronously to the search results?_

**Comment by scottkosty** [comment]
> Is there some way such that one can instantly see partial search result and rest of the result appends asynchronously to the search results?

I believe the current answer is "no". You would like analogous behavior to "less".

This might be a cool feature. I would personally want to display the last (i.e., newest) messages, in which case I suppose I would need to configure the notmuch query search to search last or something like that.

I propose to remove the tag "question" and add the tag "enhancement", unless this feature is not feasible in which case I guess a "wontfix" is appropriate.

**Comment by jindraj** [comment]
If you're loading large mailboxes I would recommend you following options:

1. using `vfolder-from-query` instead of loading mailbox specified by mailboxes command so you can be more detailed in search query.

2. look at neomutt's feature for notmuch time based queries:
```
set nm_query_window_timebase = "day"     # day, week, month or year
set nm_query_window_duration = 1         # how many of nm_query_window_timebase you want to load
set nm_query_window_current_position = 0 # timebase and duration to display.
                                         # 0 is the first timebase and duration,
                                         # 1 is next (actually previous, since we're going to the past), … 
```
After changing these configs you need to reopen the mailbox!!!

----
It's a little clumsy for use, because you need to set the variables when you want to work with time based queries. But you can create macros to make it more pleasant.

My config example:
- [bindings](https://github.com/jindraj/dotfiles/blob/master/.neomutt/neomuttrc.bindings#L76-L106) - you can replace the inline lua scripts with `<enter-command>set nm_query_window_duration+=1<enter>` etc…
- [lua script](https://github.com/jindraj/dotfiles/blob/master/.neomutt/lua/rotate_window_timebase.lua) for rotating over the time bases

**Comment by pasha009** [comment]
> My config example:
>    bindings - you can replace the inline lua scripts with <enter-command>set nm_query_window_duration+=1<enter> etc…
>    lua script for rotating over the time bases


@jindraj this looks great !
I am having a problem though.

In `rotate_window_timebase` macro the timebase doesn't appears to change if the window is empty. But when i shift vfolder window, the timebase has the expected value. Is there a way to sort this problem ?


```
# muttrc

set my_range_from = ""
set nm_query_window_timebase = "month"  
set nm_query_window_duration = 1        
set nm_query_window_current_position = 0 

macro index,pager ,< \
"<enter-command>lua-source $my_lua_scripts/rotate_tb<enter> \
<enter-command>source $my_basedir/common<enter> \
<change-vfolder>^<enter>" \
"Reset window duration"
```

```
-- $my_lua_scripts/rotate_tb

tb = mutt.get("nm_query_window_timebase")
base = { "hour", "day", "week", "month", "year" }
for k,v in pairs(base) do
  if v == tb then
    i=(k+1)%(#base+1) -- because LUA counts from 1
    if i==0 then i=1 end -- don't know if this can be done better
    break
  end
end
-- set the value
mutt.set('nm_query_window_timebase',base[i])
```

```
# $my_base_dir/common

source colorscheme/monokai.muttrc
lua-source $my_lua_scripts/compute_range

set status_chars = " *%A"
set status_format = "[ Folder: %f ] [%r%m messages%?n? (%n new)?%?d? (%d to delete)?%?t? (%t tagged)? ]%>─%?p?( %p postponed )? ${my_range_from} "
set date_format = "%d.%m.%Y %H:%M"
set index_format = "[%Z] %?X?A&-? %D  %-20.20F  %s"
```

```
-- $my_lua_scripts/compute_range

cp = mutt.get("nm_query_window_current_position")
wd = mutt.get("nm_query_window_duration")
tb = mutt.get("nm_query_window_timebase")
rf = ""

if wd > 0 then
    rf = "( from " .. cp * wd .. ".." .. (cp + 1) * wd .. " " .. tb .. " old )"
end

mutt.set("my_range_from", rf)
```

**Comment by jindraj** [comment]
> In `rotate_window_timebase` macro the timebase doesn't appears to change if the window is empty. But when i shift vfolder window, the timebase has the expected value. Is there a way to sort this problem.

Just set anywhere in the config initial value for `nm_query_window_timebase`.

**Comment by pasha009** [comment]
> Just set anywhere in the config initial value for `nm_query_window_timebase`.

I have initialized `nm_query_window_timebase`, so that is not the problem. 
I have updated previous comment with relevant code.

**Comment by flatcap** [comment]
> Is there some way such that one can instantly see partial search result
> and rest of the result appends asynchronously to the search results?

tl;dr No :-)

There are a number of problems, but it's mostly because of the code.
**The code** we inherited from Mutt is poorly designed; making it multi-threaded **now** would be disastrous.
(we're working on this, but it'll take time)

The question is, **which is the slow part of a query?**

If it's **Notmuch** -- the query is inherently time-consuming -- then there's little that NeoMutt can do.
NeoMutt sends the query to Notmuch.  Notmuch doesn't return until it has all the results.

If **NeoMutt** is slow processing the results -- **more likely** -- then perhaps we could "chunk" the results.
(Imap and Nntp do something similar).
After processing, say, 100 emails, NeoMutt would force a refresh of the display.
Some "coming soon" architectural changes would help to make this possible.
Of course there'd have to be some way to interrupt the processing.  I'm not sure how that'd work, yet.
